yang/frr-ospfd.yang

```diff
@@ -0,0 +1,957 @@
+          description
+            "Prefix SID.";
+        }
+
+        leaf no-php-flag {
+          type boolean;
+          description
+            "Don't request Penultimate Hop Popping (PHP).";
+        }
+      }
+    }
+  }
+
+  grouping ospf-list-group {
+    list neighbor {
+      key "ip";
+      description
+        "Neighbor list.";
+      leaf ip {
+        type inet:ipv4-address;
+        description
+          "Neighbor IP address.";
+      }
+
+      leaf priority {
+        type uint8;
+        description
+          "Neighbor Priority.";
+      }
+
+      leaf poll-interval {
+        type uint16 {
+          range "1..65535";
+        }
+        units "seconds";
+        description
+          "Dead Neighbor Polling interval.";
+      }
+    }
+
+    list network {
+      key "prefix";
+      description
+        "Enable routing on list of IP network.";
+      leaf prefix {
+        type inet:ipv4-prefix;
+        description
+          "Prefix on which routing needs to be enabled.";
+      }
+
+      leaf area {
+        type ospf-area-id;
+        description
+          "Area ID for this network.";
+      }
+    }
+
+    list passive-interface {
+      key "interface";
+      description
+        "Suppress routing updates on list interface.";
+      leaf interface {
+        type frr-interface:interface-ref {
+        }
+        description
+          "Suppress routing updates on an interface.";
+      }
+
+      leaf address {
+        type inet:ipv4-address;
+        description
+          "Interface address.";
+      }
+    }
+  }
+
+  grouping interface-ospf-attribute-group {
+    leaf area {
+      type ospf-area-id;
+      description
+        "Ospf area ID.";
+    }
+
+    uses authentication-group;
+
+    leaf cost {
+      type uint16 {
+        range "1..65535";
+      }
+      description
+        "Interface cost";
+    }
+
+    container dead-interval {
+      leaf interval {
+        type uint16 {
+          range "1..65535";
+        }
+      }
+
+      container minimal {
+        leaf hello-multiplier {
+          type uint8 {
+            range "1..10";
+          }
+        }
+      }
+    }
+
+    leaf hello-interval {
+      type uint16 {
+        range "1..65535";
+      }
+      description
+        "Time between HELLO packets.";
+    }
+
+    leaf retransmit-interval {
+      type uint16 {
+        range "3..65535";
+      }
+      description
+        "Time between retransmitting lost link state advertisements.";
+    }
+
+    leaf transmit-delay {
+      type uint16 {
+        range "1..65535";
+      }
+      description
+        "Link state transmit delay.";
+    }
+
+    leaf mtu-ignore {
+      type boolean;
+      description
+        "Disable MTU mismatch detection on this interface.";
+    }
+
+    leaf priority {
+      type uint8;
+      description
+        "Router priority.";
+    }
+  }
+
+  grouping interface-ospf-group {
+    list instance {
+      key "id";
```

Comment:
This list is unnecessary. The "control-plane-protocol" list (augmented by this module) already has a "name" leaf which is supposed to hold the routing instance ID.

Reply:
Since we are augmenting the interface as shown below, we need the ospf routing instance ID:
augment "/frr-interface:lib/frr-interface:interface" {
    container ospf {
      description
        "OSPF interface parameters.";
      uses interface-ospf-group;
    }

Reply:
Sorry, I mean it's the other "list instance {}" that is unnecessary. The "name" leaf under "routing/control-plane-protocols/control-plane-protocol" should be enough to distinguish different instances of a routing protocol. Example:
```
{
    "frr-routing:routing": {
        "control-plane-protocols": {
            "control-plane-protocol": {
                "type": "frr-ospfd:ospfd",
                "name": "1",
                "vrf": "default",
                "frr-ospfd:ospf": {
                   ...
                }
            },
            "control-plane-protocol": {
                "type": "frr-ospfd:ospfd",
                "name": "2",
                "vrf": "default",
                "frr-ospfd:ospf": {
                   ...
                }
            }
        }
    }
}
```

What I'm trying to say is that creating a new list of instances under the "ospf" container would be redundant.

Here, for interfaces, we should keep the list of instances, but preferably using a leafref to [1] instead of using a raw `uint16`.

[1] `/frr-routing:routing/frr-routing:control-plane-protocols/frr-routing:control-plane-protocol/name`

Reply:
Just one additional note: since `"control-plane-protocol/name"` is a string, ospfd will need to use the `routing_conf_event` hook to validate that the instance ID is a 16-bit wide unsigned integer. Other daemons like isisd use plain strings to identify routing instances and won't need any additional code-level validation.

Reply:
Removed the list.